# Telecom Customer Churn Prediction using Machine Learning

This project focuses on predicting customer churn using Machine Learning techniques on a telecom customer dataset. The workflow includes data preprocessing, exploratory data analysis, model training, and evaluation using Logistic Regression.

## Importing Required Libraries

In [19]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

## Loading the Dataset

In [2]:
import pandas as pd
df = pd.read_csv('telecomchurn_analysis.csv')
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.50,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## Dataset Information

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        5000 non-null   object 
 1   gender            5000 non-null   object 
 2   SeniorCitizen     5000 non-null   int64  
 3   Partner           5000 non-null   object 
 4   Dependents        5000 non-null   object 
 5   tenure            5000 non-null   int64  
 6   PhoneService      5000 non-null   object 
 7   MultipleLines     5000 non-null   object 
 8   InternetService   5000 non-null   object 
 9   OnlineSecurity    5000 non-null   object 
 10  OnlineBackup      5000 non-null   object 
 11  DeviceProtection  5000 non-null   object 
 12  TechSupport       5000 non-null   object 
 13  StreamingTV       5000 non-null   object 
 14  StreamingMovies   5000 non-null   object 
 15  Contract          5000 non-null   object 
 16  PaperlessBilling  5000 non-null   object 


## Missing Value Check

Checked the dataset for missing values to ensure data quality before preprocessing and model building.

In [4]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

### Observation
No missing values were found in the dataset. The data was complete and suitable for further preprocessing and model building.

## Churn Distribution Analysis

Analyzed the distribution of churn values to understand the proportion of customers who stayed and customers who left the service.

In [5]:
df['Churn'].value_counts()

Churn
No     3683
Yes    1317
Name: count, dtype: int64

### Observation
The dataset contains more non-churn customers than churn customers. Most customers stayed with the telecom service, while a smaller portion of customers left the service.

## Contract Type vs Churn Analysis

Performed cross-tab analysis between contract type and churn status to identify customer retention patterns across different contract durations.

In [6]:
pd.crosstab(df['Contract'], df['Churn'])

Churn,No,Yes
Contract,,
Month-to-month,1603,1175
One year,918,114
Two year,1162,28


### Observation

Customers with month-to-month contracts showed significantly higher churn compared to customers with one-year and two-year contracts. Longer contract durations were associated with better customer retention.

# Data Preprocessing

### Removing unnecessary columns


In [7]:
df = df.drop(['customerID'], axis=1)
df.head()

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.50,No
2,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


### Observation
The customerID column was removed because it does not contribute to predicting customer churn and acts only as a unique identifier.

# Encoding Categorical Columns
### Machine learning algorithms cannot process text-based categorical values directly.Therefore, categorical columns were converted into numerical format using Label Encoding for model training.

In [8]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

for column in df.columns:
    if df[column].dtype == 'object':
        df[column] = le.fit_transform(df[column])

df.head()

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,0,0,1,0,1,0,1,0,0,2,0,0,0,0,0,1,2,29.85,29.85,0
1,1,0,0,0,34,1,0,0,2,0,2,0,0,0,1,0,3,56.95,1889.50,0
2,1,0,0,0,2,1,0,0,2,2,0,0,0,0,0,1,3,53.85,108.15,1
3,1,0,0,0,45,0,1,0,2,0,2,2,0,0,1,0,0,42.30,1840.75,0
4,0,0,0,0,2,1,0,1,0,0,0,0,0,0,0,1,2,70.70,151.65,1


### Observation
Categorical columns were successfully converted into numerical values using Label Encoding, making the dataset suitable for machine learning model training

# Feature Selection
### The dataset was divided into input features (X) and target variable (y). Input features contain customer-related information used for prediction, while the target variable represents customer churn status.

In [9]:
X = df.drop('Churn', axis=1)

y = df['Churn']

print(X.shape)
print(y.shape)

(5000, 19)
(5000,)


### Observation
The dataset was successfully divided into input features (X) and target variable (y), preparing the data for model training and prediction.

# Train-Test Split

### The dataset was split into training and testing sets to evaluate model performance on unseen data. 80% of the data was used for training and 20% for testing.


In [11]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)
print(X_train.shape)
print(X_test.shape)

(4000, 19)
(1000, 19)


### Observation
The dataset was successfully split into training and testing sets, allowing the model to be trained on 80% of the data and tested on the remaining 20%.

# Machine Learning Model - Logistic Regression

### Logistic Regression was used to build a classification model for predicting customer churn. The model learns patterns from training data and predicts whether a customer is likely to churn or not.

In [13]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(max_iter=2000)

model.fit(X_train, y_train)

LogisticRegression(max_iter=2000)

### Observation
The Logistic Regression model was successfully trained on the customer churn dataset to identify patterns and predict customer churn behavior.

## Model Prediction
### Predictions were generated on the test dataset to evaluate how well the trained Logistic Regression model can identify customer churn.

In [14]:
y_pred = model.predict(X_test)

print(y_pred[:10])

[0 0 0 1 0 0 0 0 0 0]


### Observation
The trained Logistic Regression model successfully generated churn predictions for the test dataset, identifying customers likely to stay or leave the service.

## Model Evaluation - Accuracy Score
### The accuracy score was calculated to measure how well the Logistic Regression model predicts customer churn on unseen test data.

In [15]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)

print("Accuracy Score:", accuracy)

Accuracy Score: 0.788


### Observation
The Logistic Regression model achieved an accuracy score of 78.8%, showing good performance in predicting customer churn on unseen test data.

## Confusion Matrix
### A confusion matrix was generated to evaluate the performance of the classification model by comparing actual churn values with predicted churn values.

In [16]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred)

print(cm)

[[638  78]
 [134 150]]


### Observation
The confusion matrix showed that the Logistic Regression model correctly classified most customer churn and non-churn cases, indicating good prediction performance.

## Classification Report
### A classification report was generated to evaluate the overall performance of the Logistic Regression model using precision, recall, and F1-score metrics.

In [17]:
from sklearn.metrics import classification_report

report = classification_report(y_test, y_pred)

print(report)

              precision    recall  f1-score   support

           0       0.83      0.89      0.86       716
           1       0.66      0.53      0.59       284

    accuracy                           0.79      1000
   macro avg       0.74      0.71      0.72      1000
weighted avg       0.78      0.79      0.78      1000



### Observation
The classification report showed that the Logistic Regression model achieved good overall performance with balanced precision, recall, and F1-score values for customer churn prediction.

## Project Conclusion

This project successfully analyzed customer churn data and built a Logistic Regression model to predict customer churn behavior. Data preprocessing techniques such as removing unnecessary columns and encoding categorical variables were applied to prepare the dataset for machine learning. Exploratory analysis revealed that customers with month-to-month contracts had a higher chance of churning compared to customers with longer contracts.

The Logistic Regression model achieved an accuracy score of approximately 78.8%, showing good performance in predicting customer churn. Evaluation metrics such as the confusion matrix and classification report confirmed that the model was able to classify customer churn effectively. Overall, this project demonstrates how machine learning can help businesses identify customers at risk of leaving and improve customer retention strategies..